File: templates/generate_docx_template.js
/**
 * DOCX Generator Template for Review Papers
 * 
 * USAGE:
 *   1. npm install docx
 *   2. Replace the placeholder content below with your actual paper content
 *   3. node generate_docx_template.js
 * 
 * This template follows the review_paper_format_guide.md formatting rules:
 *   - Times New Roman throughout
 *   - 24pt Bold centered title
 *   - 3-column borderless author table
 *   - 7-column comparison table
 *   - Page numbers in footer
 *   - US Letter size (8.5 x 11 inches)
 */

const {
  Document, Packer, Paragraph, TextRun, Table, TableRow, TableCell,
  AlignmentType, HeadingLevel, BorderStyle, WidthType, ShadingType,
  VerticalAlign, ExternalHyperlink, PageNumber, Header, Footer,
  LevelFormat, PageBreak
} = require('docx');
const fs = require('fs');

// ─── CONFIGURATION ────────────────────────────────────────────────────────────
const CONFIG = {
  title: '[YOUR PAPER TITLE HERE]',
  runningTitle: '[Short Running Title]',
  outputFile: 'REVIEW_PAPER.docx',
  authors: [
    { name: 'Student A', dept: 'Department of Computer Science', uni: 'University Name, City, State', email: 'studentA@university.edu' },
    { name: 'Student B', dept: 'Department of Computer Science', uni: 'University Name, City, State', email: 'studentB@university.edu' },
    { name: 'Professor Name', designation: 'Assistant Professor', dept: 'Department of Computer Science', uni: 'University Name, City, State', email: 'professor@university.edu' },
  ]
};

// ─── BORDERS ──────────────────────────────────────────────────────────────────
const NO_BORDER = { style: BorderStyle.NONE, size: 0, color: 'FFFFFF' };
const NO_BORDERS = { top: NO_BORDER, bottom: NO_BORDER, left: NO_BORDER, right: NO_BORDER };
const TABLE_BORDER = { style: BorderStyle.SINGLE, size: 1, color: 'AAAAAA' };
const TABLE_BORDERS = { top: TABLE_BORDER, bottom: TABLE_BORDER, left: TABLE_BORDER, right: TABLE_BORDER };

// ─── HELPERS ──────────────────────────────────────────────────────────────────
const tnr = (text, opts = {}) => new TextRun({ text, font: 'Times New Roman', ...opts });

function body(text) {
  return new Paragraph({
    alignment: AlignmentType.JUSTIFIED,
    spacing: { after: 120 },
    children: [tnr(text, { size: 24 })]
  });
}

function sectionHeading(text) {
  return new Paragraph({
    spacing: { before: 280, after: 140 },
    children: [tnr(text, { size: 28, bold: true })]
  });
}

function tableCell(text, widthDxa, shade = null, bold = false, size = 20) {
  return new TableCell({
    borders: TABLE_BORDERS,
    width: { size: widthDxa, type: WidthType.DXA },
    shading: shade ? { fill: shade, type: ShadingType.CLEAR } : undefined,
    margins: { top: 60, bottom: 60, left: 100, right: 100 },
    verticalAlign: VerticalAlign.TOP,
    children: [new Paragraph({ children: [tnr(text, { size, bold })] })]
  });
}

// ─── AUTHOR BLOCK ─────────────────────────────────────────────────────────────
function makeAuthorBlock() {
  const colW = 3120;
  const makeCell = (author) => new TableCell({
    borders: NO_BORDERS,
    width: { size: colW, type: WidthType.DXA },
    margins: { top: 60, bottom: 60, left: 80, right: 80 },
    children: [
      new Paragraph({ alignment: AlignmentType.CENTER, children: [tnr(author.name, { size: 24, bold: true })] }),
      ...(author.designation ? [new Paragraph({ alignment: AlignmentType.CENTER, children: [tnr(author.designation, { size: 24 })] })] : []),
      new Paragraph({ alignment: AlignmentType.CENTER, children: [tnr(author.dept, { size: 24 })] }),
      new Paragraph({ alignment: AlignmentType.CENTER, children: [tnr(author.uni, { size: 24 })] }),
      new Paragraph({
        alignment: AlignmentType.CENTER,
        children: [new ExternalHyperlink({
          link: `mailto:${author.email}`,
          children: [new TextRun({ text: author.email, font: 'Times New Roman', size: 24, style: 'Hyperlink', underline: {} })]
        })]
      }),
    ]
  });

  return new Table({
    width: { size: 9360, type: WidthType.DXA },
    columnWidths: [colW, colW, colW],
    rows: [new TableRow({ children: CONFIG.authors.map(makeCell) })]
  });
}

// ─── CONTENT (REPLACE THESE WITH YOUR ACTUAL CONTENT) ─────────────────────────
const abstractText = '[YOUR ABSTRACT TEXT HERE - 200-300 words covering all 6 required points]';
const keywordsText = 'Keyword One, Keyword Two, Keyword Three, Keyword Four, Keyword Five';

const introParas = [
  '[Introduction paragraph 1]',
  '[Introduction paragraph 2]',
  '[Introduction paragraph 3]',
];

const objectiveText = '[Your objective paragraph here - 5-8 sentences, third person]';

// Literature review: array of 30 objects { text, citation }
const litReview = [
  // { text: 'Paragraph text here...', citation: '[1]' },
  // ... repeat for all 30 papers
];

// Comparison table: 5 rows of 7 columns
const compRows = [
  // ['1', 'Paper Title', 'Author et al. (Year)', 'Year', 'Objective', 'Methodology', 'Result'],
  // ... repeat for 5 papers
];

const conclusionParas = [
  '[Conclusion paragraph 1 - summary]',
  '[Conclusion paragraph 2 - challenges]',
  '[Conclusion paragraph 3 - forward-looking]',
];

const futureText = '[Future scope paragraph - ~150 words]';

// References: array of 30 APA-formatted strings
const references = [
  // '[1] Author, A. (Year). Title. Journal, Vol(Issue), pages.',
  // ... repeat for 30
];

// ─── BUILD DOCUMENT ───────────────────────────────────────────────────────────
const children = [
  // Title
  new Paragraph({
    alignment: AlignmentType.CENTER,
    spacing: { after: 200 },
    children: [tnr(CONFIG.title, { size: 48, bold: true })]
  }),

  // Authors
  makeAuthorBlock(),
  new Paragraph({ spacing: { after: 200 }, children: [] }),

  // Abstract
  new Paragraph({
    alignment: AlignmentType.JUSTIFIED,
    spacing: { after: 120 },
    children: [
      tnr('Abstract\u2014', { size: 24, bold: true }),
      tnr(abstractText, { size: 24 })
    ]
  }),

  // Keywords
  new Paragraph({
    alignment: AlignmentType.JUSTIFIED,
    spacing: { after: 240 },
    children: [
      tnr('Keywords\u2014', { size: 24, bold: true }),
      tnr(keywordsText, { size: 24 })
    ]
  }),

  // Introduction
  sectionHeading('I. Introduction'),
  ...introParas.map(p => body(p)),

  // Objective
  sectionHeading('II. Objective'),
  body(objectiveText),

  // Literature Review
  sectionHeading('III. Literature Review'),
  ...litReview.map(({ text, citation }) => new Paragraph({
    alignment: AlignmentType.JUSTIFIED,
    spacing: { after: 120 },
    children: [
      tnr(text, { size: 24 }),
      tnr(' ' + citation, { size: 24, bold: true })
    ]
  })),

  // Comparison Table
  sectionHeading('IV. Comparison of 30 Published Research Papers'),
  body('[Introductory paragraph explaining why comparison is needed]'),
  new Paragraph({ spacing: { after: 80 }, children: [tnr('Table 1. Comparative Summary of Selected Published Research Papers', { size: 18, bold: true })] }),
  ...(compRows.length > 0 ? [new Table({
    width: { size: 9360, type: WidthType.DXA },
    columnWidths: [500, 1500, 1300, 500, 1700, 1800, 2060],
    rows: [
      new TableRow({
        tableHeader: true,
        children: ['S.No', 'Title of the RP', 'Author Name', 'Year', 'Objective', 'Methodology', 'Conclusion / Result']
          .map((h, i) => tableCell(h, [500, 1500, 1300, 500, 1700, 1800, 2060][i], 'D5E8F0', true, 18))
      }),
      ...compRows.map(row => new TableRow({
        children: row.map((cell, i) => tableCell(cell, [500, 1500, 1300, 500, 1700, 1800, 2060][i], null, false, 18))
      }))
    ]
  })] : []),

  // Conclusion
  sectionHeading('V. Conclusion'),
  ...conclusionParas.map(p => body(p)),

  // Future Scope
  sectionHeading('VI. Future Scope'),
  body(futureText),

  // References
  sectionHeading('References'),
  ...references.map(ref => new Paragraph({
    alignment: AlignmentType.JUSTIFIED,
    spacing: { after: 100 },
    indent: { left: 360, hanging: 360 },
    children: [tnr(ref, { size: 20 })]
  }))
];

// ─── ASSEMBLE & WRITE ─────────────────────────────────────────────────────────
const doc = new Document({
  styles: {
    default: { document: { run: { font: 'Times New Roman', size: 24 } } },
    paragraphStyles: [
      { id: 'Heading1', name: 'Heading 1', basedOn: 'Normal', next: 'Normal', quickFormat: true,
        run: { size: 32, bold: true, font: 'Times New Roman' },
        paragraph: { spacing: { before: 280, after: 140 }, outlineLevel: 0 } },
    ]
  },
  sections: [{
    properties: {
      page: {
        size: { width: 12240, height: 15840 },
        margin: { top: 1440, right: 1440, bottom: 1440, left: 1440 }
      }
    },
    headers: {
      default: new Header({
        children: [new Paragraph({
          alignment: AlignmentType.RIGHT,
          border: { bottom: { style: BorderStyle.SINGLE, size: 6, color: '444444', space: 1 } },
          children: [tnr(CONFIG.runningTitle, { size: 18, italics: true })]
        })]
      })
    },
    footers: {
      default: new Footer({
        children: [new Paragraph({
          alignment: AlignmentType.CENTER,
          border: { top: { style: BorderStyle.SINGLE, size: 6, color: '444444', space: 1 } },
          children: [
            tnr('Page ', { size: 18 }),
            new TextRun({ children: [PageNumber.CURRENT], font: 'Times New Roman', size: 18 }),
            tnr(' of ', { size: 18 }),
            new TextRun({ children: [PageNumber.TOTAL_PAGES], font: 'Times New Roman', size: 18 })
          ]
        })]
      })
    },
    children
  }]
});

Packer.toBuffer(doc).then(buffer => {
  fs.writeFileSync(CONFIG.outputFile, buffer);
  console.log(`Done! Generated ${CONFIG.outputFile}`);
}).catch(err => {
  console.error('Error generating DOCX:', err);
});

File: templates/generate_docx_template.py
"""
DOCX Generator Template for Review Papers (Python)

USAGE:
  1. pip install python-docx
  2. Replace the placeholder content below with your actual paper content
  3. python generate_docx_template.py

This template follows the review_paper_format_guide.md formatting rules:
  - Times New Roman throughout
  - 24pt Bold centered title
  - 3-column borderless author table
  - 7-column comparison table
  - US Letter size (8.5 x 11 inches)
"""

import docx
from docx import Document
from docx.shared import Pt, Inches, RGBColor
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.enum.table import WD_TABLE_ALIGNMENT
from docx.oxml.ns import qn
from docx.oxml import OxmlElement

# ─── CONFIGURATION ────────────────────────────────────────────────────────────
CONFIG = {
    'title': '[YOUR PAPER TITLE HERE]',
    'outputFile': 'REVIEW_PAPER.docx',
    'authors': [
        {'name': 'Student A', 'dept': 'Department of Computer Science', 'uni': 'University Name, City, State', 'email': 'studentA@university.edu'},
        {'name': 'Student B', 'dept': 'Department of Computer Science', 'uni': 'University Name, City, State', 'email': 'studentB@university.edu'},
        {'name': 'Professor Name', 'designation': 'Assistant Professor', 'dept': 'Department of Computer Science', 'uni': 'University Name, City, State', 'email': 'professor@university.edu'},
    ]
}

# ─── CONTENT (REPLACE THESE WITH YOUR ACTUAL CONTENT) ─────────────────────────
abstractText = '[YOUR ABSTRACT TEXT HERE - 200-300 words covering all 6 required points]'
keywordsText = 'Keyword One, Keyword Two, Keyword Three, Keyword Four, Keyword Five'

introParas = [
    '[Introduction paragraph 1]',
    '[Introduction paragraph 2]',
    '[Introduction paragraph 3]',
]

objectiveText = '[Your objective paragraph here - 5-8 sentences, third person]'

litReview = [
    # {'text': 'Paragraph text here...', 'citation': '[1]'},
    # ... repeat for all 30 papers
]

compRows = [
    # ['1', 'Paper Title', 'Author et al. (Year)', 'Year', 'Objective', 'Methodology', 'Result'],
    # ... repeat for 5 papers
]

conclusionParas = [
    '[Conclusion paragraph 1 - summary]',
    '[Conclusion paragraph 2 - challenges]',
    '[Conclusion paragraph 3 - forward-looking]',
]

futureText = '[Future scope paragraph - ~150 words]'

references = [
    # '[1] Author, A. (Year). Title. Journal, Vol(Issue), pages.',
    # ... repeat for 30
]

# ─── HELPERS ──────────────────────────────────────────────────────────────────
doc = Document()

# Set US Letter size and 1-inch margins
for section in doc.sections:
    section.page_width = Inches(8.5)
    section.page_height = Inches(11)
    section.top_margin = Inches(1.0)
    section.bottom_margin = Inches(1.0)
    section.left_margin = Inches(1.0)
    section.right_margin = Inches(1.0)

def set_font(run, size=12, bold=False, italic=False):
    run.font.name = 'Times New Roman'
    run.font.size = Pt(size)
    run.bold = bold
    run.italic = italic

def add_heading(text, level=1, size=16):
    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.LEFT
    run = p.add_run(text)
    set_font(run, size=size, bold=True)
    p.paragraph_format.space_before = Pt(18)
    p.paragraph_format.space_after = Pt(6)

def add_paragraph(text, size=12, bold=False, justify=True):
    p = doc.add_paragraph()
    if justify:
        p.alignment = WD_ALIGN_PARAGRAPH.JUSTIFY
    run = p.add_run(text)
    set_font(run, size, bold)
    p.paragraph_format.space_after = Pt(6)
    return p

# ─── BUILD DOCUMENT ───────────────────────────────────────────────────────────

# Title
p = doc.add_paragraph()
p.alignment = WD_ALIGN_PARAGRAPH.CENTER
run = p.add_run(CONFIG['title'])
set_font(run, size=24, bold=True)
p.paragraph_format.space_after = Pt(24)

# Author Block
table = doc.add_table(rows=1, cols=3)
table.alignment = WD_TABLE_ALIGNMENT.CENTER
cells = table.rows[0].cells
for i, author in enumerate(CONFIG['authors']):
    cell = cells[i]
    p = cell.paragraphs[0]
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run = p.add_run(author['name'] + "\n")
    set_font(run, size=12, bold=True)

    if 'designation' in author:
        run = p.add_run(author['designation'] + "\n")
        set_font(run, size=12)
        
    run = p.add_run(author['dept'] + "\n" + author['uni'] + "\n")
    set_font(run, size=12)
    
    run = p.add_run(author['email'])
    set_font(run, size=12)
    run.font.underline = True
    run.font.color.rgb = RGBColor(0, 0, 255) # blue link

doc.add_paragraph() # spacing

# Abstract
p = doc.add_paragraph()
p.alignment = WD_ALIGN_PARAGRAPH.JUSTIFY
run_label = p.add_run("Abstract\u2014")
set_font(run_label, size=12, bold=True)
run_text = p.add_run(abstractText)
set_font(run_text, size=12)

# Keywords
p = doc.add_paragraph()
p.alignment = WD_ALIGN_PARAGRAPH.JUSTIFY
p.paragraph_format.space_before = Pt(6)
p.paragraph_format.space_after = Pt(18)
run_label = p.add_run("Keywords\u2014")
set_font(run_label, size=12, bold=True)
run_text = p.add_run(keywordsText)
set_font(run_text, size=12)

# I. Introduction
add_heading("I. Introduction")
for para in introParas:
    add_paragraph(para)

# II. Objective
add_heading("II. Objective")
add_paragraph(objectiveText)

# III. Literature Review
add_heading("III. Literature Review")
for item in litReview:
    p = add_paragraph(item['text'] + " ")
    run_cite = p.add_run(item['citation'])
    set_font(run_cite, size=12, bold=True)

# IV. Comparison
add_heading("IV. Comparison of 30 Published Research Papers")
add_paragraph("[Introductory paragraph explaining why comparison is needed]")

p_caption = doc.add_paragraph()
run_caption = p_caption.add_run("Table 1. Comparative Summary of Selected Published Research Papers")
set_font(run_caption, size=9, bold=True)

if compRows:
    comp_table = doc.add_table(rows=1, cols=7)
    comp_table.style = 'Grid Table 1 Light' # gives basic borders
    hdr_cells = comp_table.rows[0].cells
    headers = ['S.No', 'Title of the RP', 'Author Name', 'Year', 'Objective', 'Methodology', 'Conclusion / Result']
    for i, hc in enumerate(headers):
        hdr_cells[i].text = hc
        set_font(hdr_cells[i].paragraphs[0].runs[0], size=10, bold=True)
        
    for row_data in compRows:
        row_cells = comp_table.add_row().cells
        for i, val in enumerate(row_data):
            row_cells[i].text = val
            set_font(row_cells[i].paragraphs[0].runs[0], size=10)

# V. Conclusion
add_heading("V. Conclusion")
for para in conclusionParas:
    add_paragraph(para)

# VI. Future Scope
add_heading("VI. Future Scope")
add_paragraph(futureText)

# References
add_heading("References")
for ref in references:
    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.JUSTIFY
    p.paragraph_format.left_indent = Inches(0.5)
    p.paragraph_format.first_line_indent = Inches(-0.5)
    run = p.add_run(ref)
    set_font(run, size=10)

doc.save(CONFIG['outputFile'])
print(f"Done! Generated {CONFIG['outputFile']}")
